### PR TITLE
perf: add Services::get()

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
- add `Services::get()` method to get the shared instance fast.
- `service()` uses `Services::get()`

```
 Test 			Time 	Memory
services::timer() 	0.1337 	0 Bytes
services::get('timer') 	0.0406 	0 Bytes
service('timer') 	0.0528 	0 Bytes
```

```php
<?php

namespace App\Controllers;

use CodeIgniter\Debug\Iterator;
use Config\Services;

class Home extends BaseController
{
    public function index(): string
    {
        $iterator = new Iterator();

        Services::timer();

        $iterator->add('Services::timer()', static function () {
            $timer = Services::timer();
        });

        $iterator->add("Services::get('timer')", static function () {
            $timer = Services::get('timer');
        });

        $iterator->add("service('timer')", static function () {
            $timer = service('timer');
        });

        return $iterator->run(300000);
    }
}
```

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by lonnieezell — APPROVED** [review]

**Review by ddevsr — APPROVED** [review]

File: system/Config/BaseService.php [hunk 1]
```diff
@@ -177,6 +185,44 @@ class BaseService
...
     private static array $serviceNames = [];
 
+    /**
+     * Simple method to get an entry fast.
+     *
+     * @param string $key Identifier of the entry to look for.
+     *
+     * @return mixed Entry.
+     */
+    public static function get(string $key): mixed
```
Comment (paulbalandan) [inline system/Config/BaseService.php:195]: Applicable review until `override()`.

`mixed` should be limited to `object`, as the static array prop `$instances` accepts `object` as value. I don't think we need to accept anything to be set. Accepting `mixed` may lead to future misuses.

File: system/Config/BaseService.php [hunk 2]
```diff
@@ -177,6 +185,44 @@ class BaseService
...
+    {
+        return static::$instances[$key] ?? static::__callStatic($key, []);
+    }
+
+    /**
+     * Sets an entry.
+     *
+     * @param string $key   Identifier of the entry.
+     * @param mixed  $value Normally an object.
```
Comment (paulbalandan) [inline system/Config/BaseService.php:204]: See comment before.

File: system/Config/BaseService.php [hunk 3]
```diff
@@ -177,6 +185,44 @@ class BaseService
...
+     */
+    public static function set(string $key, mixed $value): void
+    {
+        if (isset(static::$instances[$key])) {
+            throw new InvalidArgumentException('The Entry for "' . $key . '" is already set.');
```
Comment (paulbalandan) [inline system/Config/BaseService.php:209]:
```suggestion
            throw new InvalidArgumentException('The entry for "' . $key . '" is already set.');
```

File: system/Config/BaseService.php [hunk 4]
```diff
@@ -177,6 +185,44 @@ class BaseService
...
+        }
+
+        static::$instances[$key] = $value;
+    }
+
+    /**
+     * Overrides an existing entry.
+     *
+     * @param string $key   Identifier of the entry.
+     * @param mixed  $value Normally an object.
```
Comment (paulbalandan) [inline system/Config/BaseService.php:219]: See comment above

File: system/Config/BaseService.php (hunk 1 above)
Comment (kenjis) [reply system/Config/BaseService.php]: Changed to `?object`.

File: system/Config/BaseService.php (hunk 3 above)
Comment (kenjis) [reply system/Config/BaseService.php]: Fixed.

**Review by paulbalandan — APPROVED** [review]

**Comment by lonnieezell** [comment]
The more I think about this the more confused I am about the performance issue we're seeing. All methods are static and should be called before `__callStatic` if the case matches, shouldn't it? So that should be higher performance than either `get()` or `__callStatic`.

**Comment by kenjis** [comment]
The current Services is not that simple.
`__callStatic` is called in most cases.
Because `Config\Services` extends `BaseService`.
https://github.com/codeigniter4/CodeIgniter4/blob/fd28a3fb5313f64d0be9a523cdc043986a52fd6a/app/Config/Services.php#L20

and BaseService has only `autoloader` and `locator`.
```console
$ grep function system/Config/BaseService.php 
    public static function get(string $key): ?object
    public static function set(string $key, object $value): void
    public static function override(string $key, object $value): void
    protected static function getSharedInstance(string $key, ...$params)
    public static function autoloader(bool $getShared = true)
    public static function locator(bool $getShared = true)
    public static function __callStatic(string $name, array $arguments)
    public static function serviceExists(string $name): ?string
    public static function reset(bool $initAutoloader = true)
    public static function resetSingle(string $name)
    public static function injectMock(string $name, $mock)
    protected static function buildServicesCache(): void
```

**Comment by lonnieezell** [comment]
Yeah, I guess the only ones that would hit directly would be in `app/Config/Services`. That's too bad.

**Comment by neznaika0** [comment]
How do I work with the magic of services? The analyzer swears at the returned type. It looks sad.
```php
// some code
$this->request = service('request');
```
Errors:
```diff
3) tests/system/Honeypot/HoneypotTest.php:37

    ---------- begin diff ----------
@@ @@
     /**
      * @var CLIRequest|IncomingRequest
      */
-    private $request;
+    private ?object $request = null;

     private Response $response;
    ----------- end diff -----------

Applied rules:
 * TypedPropertyFromAssignsRector
```

**Pull request merged**